### ISSUE TITLE
Helmet at nodejs project. Content Security Policy directive

### ISSUE THREAD

**Issue by eugenestepanyuk** [helmetjs/helmet#401]
Hi! I'm trying to add to the electronic signature project helmet. The widget is connected via iframe.
`.use(helmet())`
And after that there is a mistake of this kind
![1](https://user-images.githubusercontent.com/48010641/220558932-addcf616-316a-4743-8953-32019c5b4ab9.jpg)

After that I tried to add directives to contentSecurityPolicy:
```
.use(helmet())
.use(
    helmet.contentSecurityPolicy({
    directives: {
       ...helmet.contentSecurityPolicy.getDefaultDirectives(),
       "script-src": ["'self'", "'unsafe-inline'", "http://localhost:80/"],
       "script-src-attr": ["'none'", "'unsafe-inline'", "http://localhost:80/"],
       "connect-src": ["'self'", "http://localhost:80/"],
       },
   })
)
```
But such a mistake arises, the solution of which I did not find, maybe there will be some advice
![2](https://user-images.githubusercontent.com/48010641/220559044-b45856c7-8598-40d8-ae68-770cfe76aeda.jpg)

I use the script directly on the HTML page and read the information that helmet does not like it and I need to put the script into the module and connect it. Right?
`<script type="module" src="..."></script>`
I did it, but the error remains.
![3](https://user-images.githubusercontent.com/48010641/220559353-a5e58a8c-1642-4887-b0ff-0889c266e39a.jpg)

I tried many options, the question arose, how safe is it if you completely disable contentSecurityPolicy?

I will be grateful in helping this problem

**Comment by EvanHahn** [comment]
It looks like the `Content-Security-Policy` header is not being updated after you make changes. Can you do a hard refresh in your browser?

**Comment by eugenestepanyuk** [comment]
Yes, sure, I tried a hard refresh, the mistake remains

**Comment by eugenestepanyuk** [comment]
Content-Security-Policy: default-src 'self';base-uri 'self';font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests;connect-src 'self' http://localhost:3030/;frame-src 'self' https://eu.msn.com.ua/

**Comment by EvanHahn** [comment]
I'm not sure if this is a problem with Content Security Policy or with a different part of Helmet.

Do you see different behavior between these two snippets?

```js
// Just use Helmet's Content-Security-Policy middleware.
// `helmet()` should not be used elsewhere.
app.use(
  helmet.contentSecurityPolicy({
    directives: {
      ...helmet.contentSecurityPolicy.getDefaultDirectives(),
      "connect-src": ["'self'", "http://localhost:3030/"],
      "frame-src": ["'self'", "https://eu.msn.com.ua/"],
    },
  }),
);
```

```js
// Use all of Helmet, but configure the Content-Security-Policy middleware.
app.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        ...helmet.contentSecurityPolicy.getDefaultDirectives(),
        "connect-src": ["'self'", "http://localhost:3030/"],
        "frame-src": ["'self'", "https://eu.msn.com.ua/"],
      },
    },
  }),
);
```

**Comment by eugenestepanyuk** [comment]
found this warning in Nerwork tab
Request URL: https://eu.msn.com.ua/sign-widget/v20200811/?address=http://localhost&formType=1&debug=false 
![1](https://user-images.githubusercontent.com/48010641/221114864-caf2b60f-b34d-45fe-8e37-980971014376.jpg)

Could this be the reason for blocking the widget?
I changed the code to
```
.use(
        helmet({
          contentSecurityPolicy: {
            directives: {
              ...helmet.contentSecurityPolicy.getDefaultDirectives(),
              "connect-src": ["'self'", "http://localhost:3030/"],
              "frame-src": ["'self'", "https://eu.msn.com.ua/"],
            },
          },
          crossOriginResourcePolicy: true
        })
      )
```

**Comment by EvanHahn** [comment]
You're embedding an image from another origin, triggering the document's "cross-origin embedder policy". This policy, controlled by the `Cross-Origin-Embedder-Policy` header, has three modes:

1. `unsafe-none`, the default value. Your page can load cross-origin resources without permission from the remote origin.

   For example, your page on `example.com` would be able to load `<img src="https://example.com/foo.png">` with no issues.
2. `require-corp`, Helmet's default value. Your page can load cross-origin resources, but the remote origin has to give explicit permission. It can do this with the [`Cross-Origin-Resource-Policy` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Resource-Policy) or by using CORS.

   For example, your page on `example.com` can load `<img src="https://example.com/foo.png">` _only if_ `uilogos.co` allows it. They'd need to do some work to support this.
3. `credentialless`. Your page can load cross-origin resources, but they can't include credentials like cookies. This isn't supported by all browsers.

I would recommend trying the `credentialless` policy to see if that works for you.

```js
app.use(
  helmet({
    // ...
    crossOriginEmbedderPolicy: { policy: 'credentialless' }
  })
)
```

**Comment by eugenestepanyuk** [comment]
adding `credentialless` policy didn't help, the warning remains for some reason

**Comment by EvanHahn** [comment]
What if you disable it?

```
app.use(
  helmet({
    // ...
    crossOriginEmbedderPolicy: false
  })
)
```

**Comment by eugenestepanyuk** [comment]
didn't help either

**Comment by EvanHahn** [comment]
To be clear: did you set the `crossorigin` attribute on the iframe?

**Comment by eugenestepanyuk** [comment]
This is how the iframe element is created in the signature library file
```
var iframe = document.createElement("iframe");
iframe.setAttribute("src", src + srcParams);
iframe.setAttribute("id", id);
iframe.setAttribute("frameborder", "0");
iframe.setAttribute("allowtransparency", "true");
iframe.setAttribute("width", "100%");
iframe.setAttribute("height", "100%");
iframe.setAttribute("crossorigin", "anonymous");
document.getElementById(parentId).appendChild(iframe);
```

I added the line `iframe.setAttribute("crossorigin", "anonymous");` but it did not help
Access is denied, as I understand it, on the line adding an iframe element to the page `document.getElementById(parentId).appendChild(iframe);`

**Comment by eugenestepanyuk** [comment]
The issue still persists, is there any other solution?

**Comment by EvanHahn** [comment]
Sorry for my slow response.

I don't know why this is happening, but some part of Helmet seems to be causing it.

The top-level `helmet` function is a wrapper around 14 smaller middlewares. I would start by disabling all of them and slowly re-introducing them, one by one,

For example, start with this:

```js
app.use(helmet({
  contentSecurityPolicy: false,
  crossOriginEmbedderPolicy: false,
  crossOriginOpenerPolicy: false,
  crossOriginResourcePolicy: false,
  dnsPrefetchControl: false,
  frameguard: false,
  hidePoweredBy: false,
  hsts: false,
  ieNoOpen: false,
  noSniff: false,
  originAgentCluster: false,
  permittedCrossDomainPolicies: false,
  referrerPolicy: false,
  xssFilter: false,
});
```

This won't do anything at all!

Then _just_ try reintroducing CSP:

```js
app.use(helmet({
  contentSecurityPolicy: { /* ... */ },
  crossOriginEmbedderPolicy: false,
  crossOriginOpenerPolicy: false,
  // ...
});
```

Then try introducing the next one:

```js
app.use(helmet({
  contentSecurityPolicy: { /* ... */ },
  crossOriginEmbedderPolicy: true,
  crossOriginOpenerPolicy: false,
  // ...
});
```

Eventually, I hope this will help you find the culprit.

**Comment by eugenestepanyuk** [comment]
I tested middlewares
```
.use(
        helmet({
          contentSecurityPolicy: {
            directives: {
              ...helmet.contentSecurityPolicy.getDefaultDirectives(),
              "connect-src": ["'self'", "http://localhost:3030/"],
              "frame-src": ["'self'", "https://eu.iit.com.ua/"],
            },
          },
          crossOriginEmbedderPolicy: false,
          crossOriginOpenerPolicy: true,
          crossOriginResourcePolicy: true,
          dnsPrefetchControl: true,
          frameguard: true,
          hidePoweredBy: true,
          hsts: true,
          ieNoOpen: true,
          noSniff: true,
          originAgentCluster: true,
          permittedCrossDomainPolicies: true,
          referrerPolicy: true,
          xssFilter: true,
        })
      )
```
in this format the widget opens, mby problem in` crossOriginEmbedderPolicy`, `{ policy: "credentialless" }` dosen't help

**Comment by eugenestepanyuk** [comment]
I tested middlewares
```
.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        ...helmet.contentSecurityPolicy.getDefaultDirectives(),
        "connect-src": ["'self'", "http://localhost:3030/"],
        "frame-src": ["'self'", "https://eu.msn.com.ua/"],
      },
    },
    crossOriginEmbedderPolicy: false,
    crossOriginOpenerPolicy: true,
    crossOriginResourcePolicy: true,
    dnsPrefetchControl: true,
    frameguard: true,
    hidePoweredBy: true,
    hsts: true,
    ieNoOpen: true,
    noSniff: true,
    originAgentCluster: true,
    permittedCrossDomainPolicies: true,
    referrerPolicy: true,
    xssFilter: true,
  })
)
```
in this format the widget opens, mby problem in `crossOriginEmbedderPolicy`, `{ policy: "credentialless" }` dosen't help

**Comment by EvanHahn** [comment]
In that case, it sounds like disabling that one middleware solves your problem. I would recommend doing that.

Does that help?

**Comment by eugenestepanyuk** [comment]
yes, with `crossOriginEmbedderPolicy: false` the widget works.
Please tell me, will this affect safety?

**Comment by eugenestepanyuk** [comment]
yes, with `crossOriginEmbedderPolicy: false` the widget works.
Please tell me, will this affect safety? I read, this may leave your application vulnerable to certain types of attacks
And read info
`If you are experiencing an access error for the iframe widget when crossOriginEmbedderPolicy is set to true, it is likely because the iframe is trying to access a resource that is restricted by the policy. You will need to ensure that the iframe is not trying to access any restricted resources, or you can consider adding a crossOriginOpenerPolicy header to your server response, which can provide a more lenient policy for cross-origin iframe loading.`

What do you say about this?

**Comment by eugenestepanyuk** [comment]
yes, with `crossOriginEmbedderPolicy: false` the widget works.
Please tell me, will this affect safety? I read, this may leave my app vulnerable to certain types of attacks.

And read info
`If you are experiencing an access error for the iframe widget when crossOriginEmbedderPolicy is set to true, it is likely because the iframe is trying to access a resource that is restricted by the policy. You will need to ensure that the iframe is not trying to access any restricted resources, or you can consider adding a crossOriginOpenerPolicy header to your server response, which can provide a more lenient policy for cross-origin iframe loading.`

What do you say about this?

**Comment by eugenestepanyuk** [comment]
I tried set headers
```
crossOriginResourcePolicy: { policy: "same-origin" },
crossOriginOpenerPolicy: { policy: "same-origin"}
```
but that doesn't help either

**Comment by EvanHahn** [comment]
_tl;dr: You're probably safe to disable it._

First, there's the _embedder_ and the _resource_. You control the embedder—it's your website. You don't control the resource, `en.msn.com.ua`, which is what you're putting in an iframe.

`Cross-Origin-Embedder-Policy` is a header for the embedder. It says how the browser should embed cross-origin resources.

If it's set to `unsafe-none` (or is unset), the browser can embed whatever resource it wants. If it's set to `require-corp`, the browser can only embed resources that support CORS or set the `Cross-Origin-Resource-Policy` header. (There's a third option, `credentialless`, which [doesn't work in all browsers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy#browser_compatibility).)

There are some side-channel vulnerabilities when embedding an iframe in the same address space as your site. An iframe on another site might be able to read the memory of _your_ site! That's bad! But there's good news: modern browsers protect against this even without any special headers. You need these headers for certain advanced things, like [SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer), but most people don't need this. Because of this, I plan to disable this header (by default) in future versions of Helmet.

As you might expect, you have no control over `en.msn.com.ua`, so you can't make them set any headers. That's why you're seeing those errors despite setting the `Cross-Origin-Resource-Policy` header yourself: the error is coming from the resource and has nothing to do with the header you're setting.

Given that you don't control the resource, you have a few options:

1. Disable the `Cross-Origin-Embedder-Policy` header. This will let you embed whatever resources you want.
1. Get `en.msn.com.ua` to set the `Cross-Origin-Resource-Policy` header. This probably involves emailing someone and hoping they'll make your change.
1. Give up on this entirely, and stop trying to embed the iframe.

Given all that, I would recommend the first option—just disable the header on your site.

I'm going to close this issue because I think I've resolved your problem.

**Comment by eugenestepanyuk** [comment]
thank you for your great help!
